modules/ml/src/main/java/org/apache/ignite/ml/composition/stacking/StackedDatasetTrainer.java

```diff
@@ -0,0 +1,410 @@
+        this.aggregatorTrainer = aggregatorTrainer;
+
+        return this;
+    }
+
+    /**
+     * Specify binary operator used to merge submodels outputs to one.
+     *
+     * @param merger Binary operator used to merge submodels outputs to one.
+     * @return This object.
+     */
+    public StackedDatasetTrainer<IS, IA, O, AM, L> withAggregatorInputMerger(IgniteBinaryOperator<IA> merger) {
+        this.aggregatingInputMerger = merger;
+
+        return this;
+    }
+
+    /**
+     * Adds submodel trainer along with converters needed on training and inference stages.
+     *
+     * @param trainer Submodel trainer.
+     * @return This object.
+     */
+    public <M1 extends Model<IS, IA>> StackedDatasetTrainer<IS, IA, O, AM, L> addTrainer(
+        DatasetTrainer<M1, L> trainer) {
+        submodelsTrainers.add(new DatasetTrainer<Model<IS, IA>, L>() {
+            /** {@inheritDoc} */
+            @Override public <K, V> Model<IS, IA> fit(DatasetBuilder<K, V> datasetBuilder,
+                IgniteBiFunction<K, V, Vector> featureExtractor, IgniteBiFunction<K, V, L> lbExtractor) {
+                return trainer.fit(datasetBuilder, featureExtractor, lbExtractor);
+            }
+
+            /** {@inheritDoc} */
+            @Override protected boolean checkState(Model<IS, IA> mdl) {
+                return true;
+            }
+
+            /** {@inheritDoc} */
+            @Override protected <K, V> Model<IS, IA> updateModel(Model<IS, IA> mdl, DatasetBuilder<K, V> datasetBuilder,
+                IgniteBiFunction<K, V, Vector> featureExtractor, IgniteBiFunction<K, V, L> lbExtractor) {
+                return null;
+            }
+
+            /** {@inheritDoc} */
+            @Override public <K, V> Model<IS, IA> update(Model<IS, IA> mdl, DatasetBuilder<K, V> datasetBuilder,
+                IgniteBiFunction<K, V, Vector> featureExtractor, IgniteBiFunction<K, V, L> lbExtractor) {
+                DatasetTrainer<Model<IS, IA>, L> trainer1 = (DatasetTrainer<Model<IS, IA>, L>)trainer;
+                return trainer1.update(mdl, datasetBuilder, featureExtractor, lbExtractor);
+            }
+        });
+
+        return this;
+    }
+
+    /** {@inheritDoc} */
+    @Override public <K, V> StackedModel<IS, IA, O, AM> fit(DatasetBuilder<K, V> datasetBuilder,
+        IgniteBiFunction<K, V, Vector> featureExtractor,
+        IgniteBiFunction<K, V, L> lbExtractor) {
+        return runOnSubmodels(
+            ensemble ->
+                ensemble.stream()
+                    .map(tc -> (IgniteSupplier<Model<IS, IA>>)(() -> tc.fit(datasetBuilder, featureExtractor, lbExtractor)))
+                    .collect(Collectors.toList()),
+            (at, extr) -> at.fit(datasetBuilder, extr, lbExtractor),
+            featureExtractor
+        );
+    }
+
+    /** {@inheritDoc} */
+    @Override protected <K, V> StackedModel<IS, IA, O, AM> updateModel(StackedModel<IS, IA, O, AM> mdl,
```

Comment:
Hmm... We will avoid boilerplate, but seems like it will make code more error prone. Developer can forget to override this method for trainer which potentially supports updating and get an error while trying to update this model in the future whereas keeping it abstract forces developer to think if this trainer supports update and insert `NotImplementedException` more cautiously.